src/node/server.ts

```diff
@@ -513,6 +525,51 @@ export class MainServer extends Server {
 		}
 	}
 
+	protected async handleSSHSocket(socket: net.Socket, parsedUrl: url.UrlWithParsedQuery): Promise<void> {
+		this.heartbeat();
+
+		// At some point code-server should ensure some ssh client is available
+		// instead of relying on 22.
+		const sshSocket = net.connect(22, "localhost");
```

Comment:
Directly piping the data won't work since we're using the `Sender` and `Receiver` classes from `ws` to transform back and forth between websocket format and ssh format. I'm not _super_ familiar with the `Stream` api, but I think that I would still need to write this code in a wrapper `Stream` class to do this transformation if I used pipe.

Reply:
If the SSH client sends some bytes to the web socket don't we just get those exact same bytes on the other end? It seems odd we'd have to worry about the format of that data since we just want to proxy it.

Reply:
If we were using a `ws` `WebSocket`, it'd be fine. The problem is we're using an `http.Socket` which doesn't strip out all the websocket-y data, so that's what `Sender` and `Receiver` are doing here. I couldn't find public method to upgrade an `http.Socket` to a `WebSocket`, the only public API methods for that have you spawning a new `http.Server`.

Reply:
Ahh I think I see where I went wrong. I assumed Node would just handle all the necessary transformations behind since we got the socket from the `upgrade` event but I assume it doesn't actually do that?

Reply:
Do you think it would make sense to decode the web socket frames ourselves? If it's a lot of work then using `ws` makes sense but if we could avoid a dependency just for decoding/encoding these frames it might be ideal.

Reply:
Also I wonder if we can utilize `WebSocketNodeSocket` in some way since it appears to handle the web socket frames. Except I think it might be encoding its own structure into the data as well (`VSBuffer`) so that might not work

Reply:
That was my finding with `WebSocketNodeSocket` as well, looks like there's a lot of custom protocol stuff in there that won't jive well with ssh, and would require as much or more code to transform back to normal.

I could copy out the stuff I need from `ws`, but I was falling more in @sr229's camp of trying to minimize the amount of custom code here. `ws` is also a really small addition (only 1 subdep) and it's one of the most heavily used and scrutinized libraries out there, so it seems like a safe bet to me.

Happy to do whatever the preferred approach is!

Reply:
Yeah that's definitely fair. I'm cool with that, then.